Tetris: the rotation of figures has been completed and implemented, n…

gh215 opened the pull request:
…ew figures have been added, collision and fixation of figures on the playing field have been implemented

gh215 committed:
Tetris: the rotation of figures has been completed and implemented, new figures have been added, collision and fixation of figures on the playing field have been implemented

gh215 committed:
Tetris: localization hotfix

amavrin commented on Tetris/Tetris/Figure.cpp:
```diff
@@ -1,17 +1,173 @@
+				int newY = newPosition.y + row;
+
+				// Проверяем выход за границы поля
+				if (newX < 1 || newX >= SCREEN_WIDTH / 2 - 1 || newY >= SCREEN_HEIGHT - 1)
```
Я бы предложил тут сделать специальную функцию в Screen (`isVisible(Point p)`, например), чтобы нам не работать в Figure с низкоуровневыми понятиями

amavrin commented on Tetris/Tetris/Figure.cpp:
```diff
@@ -1,17 +1,173 @@
+	}
+
+	// Проверяем возможность поворота
+	for (int row = 0; row < rotatedShape.size(); row++)
```
Ну, не. Это дублирование кода между этой функцией и canMove. Нужно вынести код проверки столкновения в отдельную функцию, и использовать её в canMove и canRotate.

amavrin commented on Tetris/Tetris/Figure.cpp:
```diff
@@ -1,17 +1,173 @@
+bool Figure::checkCollision(vector<vector<bool>>& field)
+{
+	// Проверяем каждую ячейку фигуры
+	for (int row = 0; row < shape.size(); row++)
```
Ещё одно дублирование. Не, не, не.

amavrin commented on Tetris/Tetris/Figure.cpp:
```diff
@@ -1,17 +1,173 @@
+void Figure::placeFigure()
+{
+	// Проходим по всем ячейкам фигуры
+	for (size_t row = 0; row < shape.size(); row++)
```
Ой, ещё дублирование. Ты идёшь на рекорд

amavrin commented on Tetris/Tetris/Figure.cpp:
```diff
@@ -1,17 +1,173 @@
+    return rotated;
+}
+
+vector<vector<bool>> Figure::rotateCounterClockwise(vector<vector<bool>>& shape)
```
Ну что это - две отдельные функции rotateCounterClockwise и rotateClockwise.
Опять дублирование кода

amavrin commented on Tetris/Tetris/Game.cpp:
```diff
@@ -1,29 +1,47 @@
-
 void Game::draw(Screen& field)
 {
+    for (int y = 0; y < SCREEN_HEIGHT; y++)
```
Не могу согласиться. Отрисовкой двумерных массивов у нас занимается Screen::putMatrix()

amavrin reviewed (commented):
немного поревьюил, больше сейчас не успеваю. Досмотрю потом

gh215 committed:
Tetris: The following have been added or implemented:

1. Redo duplicate code
2. Work with ONLY logical coordinates everywhere except Screen
3. Create a separate Heap class that will store the actual squares
4. Reset the keys pressed after one game cycle has passed.
5. Realise accelerated falling of pieces when pressing the down key or the spacebar, if we want to ‘switch off control’ on a piece and it falls down until it hits another piece or the field
6. Drag spawnFigure into Game.spp
7. DrawBorders in Game.cpp
8. In Screen should be DrawRect (for fields)

gh215 committed:
Tetris: Made some inprovements in code, the removal of a completely filled line has been implemented

gh215 committed:
Tetris: minor fixes
